Login: Refactor handling of redirect url

stephanethomas opened the pull request:
This pull request, together with patch D5959-code, updates the handling of the `redirect_to` parameter that can be passed to the new [`Login` page](http://calypso.localhost:3000/log-in). More specifically, it moves the sanitization and validation of the redirect url to the API in order to make it more robust. It also refactors how we store this url as well as the remember me flag in the global state tree, and makes it more consistent. Finally, it adds two Tracks events that were missing, and cleans some code.
 
#### Testing instructions

1. Apply patch D5959-code
2. Point `wordpress.com`to your sandbox
3. Run `git checkout update/login-redirect` and start your server
4. Open http://calypso.localhost:3000/log-in in incognito mode
5. Check that you can still log in, and are redirected to the [`Reader` page](http://calypso.localhost:3000/)
6. Open http://calypso.localhost:3000/log-in?redirect_to=https://www.yahoo.fr in incognito mode
7. Check that you can still log in, and are still redirected to the `Reader` page
8. Open http://calypso.localhost:3000/log-in?redirect_to=/me in incognito mode
9. Check that you can still log in, and are redirected to http://calypso.localhost:3000/me
10. Open http://calypso.localhost:3000/log-in?redirect_to=https%3A%2F%2Fwordpress.com%2Fplans in incognito mode
11. Check that you can still log in, and are redirected to https://wordpress.com/plans
12. Open http://calypso.localhost:3000/log-in?redirect_to=https://haiku.blog/ in incognito mode
13. Check that you can still log in, and are redirected to https://haiku.blog/
14. Leave a comment there, it will make its author happy

Please repeat those steps with accounts that have **2FA enabled** or by **login with Google** as well.
 
#### Additional notes

We should probably still honor the `redirect_to` parameter when the user accesses the new `Login` page while already logged in. This is something that can be addressed in another pull request, unless someone object.

 
#### Reviews
 
- [x] Code
- [x] Product
- [x] Tests

hafizrahman commented:
Products LGTM with an account without 2FA, and with an account with 2FA.

gziolo commented on client/state/login/reducer.js:
```diff
@@ -25,6 +33,21 @@ export const isRequesting = createReducer( false, {
 
+export const redirectTo = createReducer( null, {
+	[ LOGIN_REQUEST ]: () => null,
+	[ LOGIN_REQUEST_SUCCESS ]: ( state, { data } ) => get( data, 'redirect_to', state ),
```
Shouldn't it default to null or empty string?

stephanethomas replied:
That's a good point :).

gziolo commented on client/state/login/test/selectors.js:
```diff
@@ -166,41 +166,12 @@ describe( 'selectors', () => {
-		it( 'should return false if the request was successful and two-factor auth is not', () => {
-			const twoFactorEnabled = isTwoFactorEnabled( {
-				login: {
-					twoFactorAuth: {
```
I think the idea here was to validate if user_id and nonce were provided.

gziolo commented on client/state/login/selectors.js:
```diff
@@ -70,21 +62,15 @@ export const getTwoFactorPushPollInProgress = state => get( state, 'login.twoFac
-	return (
-		twoFactorAuth.user_id !== ''
-	);
+	return ! isEmpty( twoFactorAuth );
```
Why do we remove check if `user_id` is properly set?

yurynix replied:
Actually I was thinking about removing it as well, since if we got two factor enabled, we'll have this reducer populated, and if not it'll be empty, so checking for `user_id` is redundant IMO, so I agree here with @stephanethomas

stephanethomas replied:
Yes, that was my reasoning. I don't really see the point with checking that specific property if we only populate `login.twoFactorAuth` when some 2FA data was returned by the login endpoint.

gziolo commented on client/state/login/actions.js:
```diff
@@ -137,12 +137,11 @@ export const loginUser = ( usernameOrEmail, password, rememberMe ) => dispatch =
  */
-export const loginUserWithTwoFactorVerificationCode = ( two_step_code, remember_me, twoFactorAuthType ) => ( dispatch, getState ) => {
+export const loginUserWithTwoFactorVerificationCode = ( twoStepCode, twoFactorAuthType ) => ( dispatch, getState ) => {
 	dispatch( { type: TWO_FACTOR_AUTHENTICATION_LOGIN_REQUEST } );
```
I recommend this reading: http://blog.isquaredsoftware.com/2017/01/idiomatic-redux-thoughts-on-thunks-sagas-abstraction-and-reusability/. Author recommends to avoid using `getState` inside action thunk whenever it's possible. /cc @yurynix 

It's fine to leave it as it is, but I think we should always consider passing all data from components whenever possible.

yurynix replied:
I disagree, we pass those parameters ( twoStepCode, userId, etc... ) to the component to just pass it back to the action.
I think in a case like this, we should avoid the additional code involved to just pass those params around.

gziolo replied:
With the current logic you are right, but in a broad sense it's risky as Redux state can diverge from component had stored when an action was triggered.

drewblaisdell replied:
I agree with @gziolo - using `getState` in these actions creates hidden dependencies. It is worth the extra few lines of code to make the dependencies of this action clearer, I think, and is less fragile. We can do this separately, though.

yurynix replied:
Why that is less fragile? The boilerplate of moving those params around sounds more fragile than having the action creator take whatever it needs from the only source of truth (=store) we suppose to have.

drewblaisdell replied:
The source of truth is the store either way, as there is no other source of these values in the codebase. The point of contention is whether the dependencies of the function should in the function signature or pulled from the store.

As mentioned, the difference is relatively minor in this case, but if `loginUserWithTwoFactorVerificationCode` depends on some external piece of state to be mutated (i.e. if it uses `getState` expecting some value to be set in the global store), consumers of this function must be aware of its internals: reading the comment about the input/output is not sufficient for understanding the dependencies of the function, as a separate action must be executed before it can succeed. This is not an issue if those dependencies are included in the function signature.

yurynix replied:
@drewblaisdell thank you for the detailed explanation, we indeed should refactor it.
I'm sorry I made that change.

drewblaisdell replied:
@yurynix Nothing to be sorry about, I'm sure I've used `getState` in this way at least as much ;) 😭

Tug commented on client/state/login/reducer.js:
```diff
@@ -51,7 +74,17 @@ const updateTwoStepNonce = ( state, { twoStepNonce, nonceType } ) => Object.assi
-	[ LOGIN_REQUEST_SUCCESS ]: ( state, { data, rememberMe } ) => data ? { ...data, remember_me: rememberMe } : null,
+	[ LOGIN_REQUEST_SUCCESS ]: ( state, { data } ) => {
+		if ( data ) {
+			const rest = omit( data, 'redirect_to' );
```
Why omit `redirect_to` here?

stephanethomas replied:
Because this is the only data returned by the login endpoint that doesn't relate to 2FA:

```json
{
  "success": true,
  "data": {
    "redirect_to": "https:\/\/haiku.blog\/",
    "user_id": 12345,
    "two_step_supported_auth_types": ["backup", "sms", "push"],
    "two_step_nonce_backup": "obc8zkxbt1n9hk6zwbaijoa8ioi54y9h",
    "two_step_nonce_sms": "qs7qdntx5afn4dp65khc2juytzx7y32z",
    "two_step_nonce_push": "gv8n6nvws1ojrh6h4c1bacrwlk6g3pql",
    "push_web_token": "1234567890:ju5ps6nl1h0ys5umz0tszphyk1kjs7x4tlsyr4f5nbo6sm2lpdzowi9l3x2j0kf4",
    "two_step_notification_sent": "push"
  }
}
```

gziolo replied:
Taking into account all comments we discussed in this PR and concerns related to 2FA, redirect param, the way we validate if 2FA is required, we probably should update the response we send from the server. If we would keep `redirect_to` and `two_factor_auth ` separate to make it would make integration with Redux state easier.

Tug commented on client/state/login/reducer.js:
```diff
@@ -25,6 +33,21 @@ export const isRequesting = createReducer( false, {
 	[ LOGIN_REQUEST_SUCCESS ]: () => false,
 } );
 
+export const redirectTo = createReducer( null, {
```
Do we need a new reducer for this? The data we need is already available in the state. Why not query it in the selector instead? (same question for rememberMe)

gziolo replied:
It is stored in different places when we use social login, but I think selector could handle it. Good idea 👍

stephanethomas replied:
The main reason for these changes was to better separate concerns in the state tree:

```json
login: {
  isRequesting: false,
  isRequestingTwoFactorAuth: false,
  magicLogin: {
    emailAddressFormInput: '',
    isFetchingEmail: false,
    requestAuthError: null,
    requestAuthSuccess: false,
    requestEmailError: null,
    requestEmailSuccess: false,
    currentView: null
  },
  redirectTo: 'https://haiku.blog/',
  rememberMe: true,
  requestError: null,
  requestNotice: null,
  requestSuccess: true,
  twoFactorAuth: {
    user_id: 12345,
    two_step_supported_auth_types: [
      'backup',
      'sms',
      'push'
    ],
    two_step_nonce_backup: 'obc8zkxbt1n9hk6zwbaijoa8ioi54y9h',
    two_step_nonce_sms: 'qs7qdntx5afn4dp65khc2juytzx7y32z',
    two_step_nonce_push: 'gv8n6nvws1ojrh6h4c1bacrwlk6g3pql',
    push_web_token: '1234567890:ju5ps6nl1h0ys5umz0tszphyk1kjs7x4tlsyr4f5nbo6sm2lpdzowi9l3x2j0kf4',
    two_step_notification_sent: 'push'
  },
  twoFactorAuthRequestError: null,
  twoFactorAuthPushPoll: {
    inProgress: true,
    success: false
  }
},
```

I don't think `redirectTo` and `rememberMe` should be stored under `login.twoFactorAuth` since they are more general login properties, and can be used in other flows such as magic login.

I also have the feeling that we would just be hiding that complexity/mess behind some selectors otherwise :).

Tug commented:
I had a question about the change of representation of the state, otherwise the code LGTM!

fabianapsimoes commented:
> Open http://calypso.localhost:3000/log-in in incognito mode
> Check that you can still log in, and are redirected to the Reader page

When I try to log in with Google, I get the following error in a notice:

```Reducer "redirectTo" returned undefined handling "SOCIAL_LOGIN_REQUEST_SUCCESS". To ignore an action, you must explicitly return the previous state.```

If I reload, I can see that I'm actually logged in. It's just the redirect that is failing.

gziolo commented on client/state/login/reducer.js:
```diff
@@ -51,7 +57,7 @@ const updateTwoStepNonce = ( state, { twoStepNonce, nonceType } ) => Object.assi
 export const twoFactorAuth = createReducer( null, {
 	[ LOGIN_REQUEST ]: () => null,
-	[ LOGIN_REQUEST_SUCCESS ]: ( state, { data, rememberMe } ) => data ? { ...data, remember_me: rememberMe } : null,
+	[ LOGIN_REQUEST_SUCCESS ]: ( state, { data, rememberMe } ) => data ? { ...data } : null,
```
I don't think we need to copy data here as it comes from request so it's a different object than the one we store.

gziolo commented on client/blocks/login/index.jsx:
```diff
@@ -69,22 +70,16 @@ class Login extends Component {
-			newHref = window.location.origin;
-		}
+		// Redirects to / if no redirect url is available
+		const url = redirectTo ? redirectTo : window.location.origin;
```
You can set default value when destructuring props:
```js
const { redirectTo: window.location.href } = this.props;
```

stephanethomas replied:
Default values only work with [`undefined` properties](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Default_values_2). This wouldn't work in that case since `redirectTo` is always provided, even if it can be `null`.

gziolo commented on client/state/login/actions.js:
```diff
@@ -195,11 +195,15 @@ export const loginSocialUser = ( service, token ) => dispatch => {
+		.then( ( response ) => {
+			dispatch( {
+				type: SOCIAL_LOGIN_REQUEST_SUCCESS,
+				redirectTo: get( response, 'body.data.redirect_to' ),
```
Should we default to `null` here?

yurynix replied:
Maybe `https://WordPress.com` is a good default here

gziolo replied:
It's using `window.location.origin` when redirecting if we have empty value stored in Redux. It's better because it works with all different domains we use.

yurynix commented on client/state/login/reducer.js:
```diff
@@ -25,6 +33,21 @@ export const isRequesting = createReducer( false, {
+	[ LOGIN_REQUEST_SUCCESS ]: ( state, { data } ) => get( data, 'redirect_to', state ),
+	[ LOGIN_REQUEST_FAILURE ]: () => null,
+	[ SOCIAL_LOGIN_REQUEST ]: () => null,
+	[ SOCIAL_LOGIN_REQUEST_SUCCESS ]: ( state, action ) => action.redirectTo,
```
you use in other places `( state, { data } )`, why not here as well `( state, { redirectTo } )` ?

stephanethomas replied:
Haha, you bet I tried that! But then variable shadowing bit me.

yurynix commented on client/blocks/login/two-factor-authentication/verification-code-form.jsx:
```diff
@@ -61,14 +60,18 @@ class VerificationCodeForm extends Component {
+		this.props.recordTracksEvent( 'calypso_two_factor_verification_code_submit' );
+
+		this.props.loginUserWithTwoFactorVerificationCode( twoStepCode, twoFactorAuthType ).then( () => {
+			this.props.recordTracksEvent( 'calypso_two_factor_verification_code_success' );
```
nitpick: that's a bit weird:
`this.props.recordTracksEvent`
but then
`onSuccess`
either we access both with `this.props` or `const { onSuccess, recordTracksEvent } = this.props;`

stephanethomas replied:
I would love to, but the problem is again variable shadowing :(.

yurynix replied:
so why not `this.props.` both? it's just weird =)

yurynix replied:
Just wondering, not implying we need to change that

stephanethomas replied:
To save a few characters :).

This makes the call to `loginUserWithTwoFactorVerificationCode()` a little bit easier to read.

gziolo commented:
I don't think there are any blockers, just a couple of small places need some clarification.

Good job tackling this refactoring and ensuring we can safely use `redirect_to` param 👍

stephanethomas commented:
Thanks for your feedback everyone!

I've addressed the comments from [Greg](https://github.com/Automattic/wp-calypso/pull/14967#discussion_r121629336) and [Fabiana](https://github.com/Automattic/wp-calypso/pull/14967#issuecomment-308072924), and rebased. Hopefully I've also clarified several points that were brought up in other comments. This is ready for review again.

drewblaisdell commented:
The code looks good, and I agree with the rationale behind your comments. However, we need to update this to support the push endpoint, as currently we do not sanitize `redirect_to` through the endpoint we poll to see if the user accepted the notification.
